sources/packages/backend/apps/workers/src/controllers/assessment/assessment.controller.ts

```diff
@@ -231,10 +251,36 @@ export class AssessmentController {
         jobLogger.error(message);
         return job.error(ASSESSMENT_NOT_FOUND, message);
       }
-      await this.studentAssessmentService.updateAssessmentStatusAndSaveWorkflowData(
-        job.variables.assessmentId,
-        job.variables.workflowData,
-      );
+      // The updateAssessmentStatusAndSaveWorkflowData and assessImpactedApplicationReassessmentNeeded are executed in the same transaction
+      // to force then to be successfully executed together or to fail together. In this way the worker can be safely retried from Camunda.
+      await this.dataSource.transaction(async (entityManager) => {
+        // Update status and workflow data ensuring that it will be updated only for the first time to ensure the worker idempotency.
+        const updated =
+          await this.studentAssessmentService.updateAssessmentStatusAndSaveWorkflowData(
+            job.variables.assessmentId,
+            job.variables.workflowData,
+            entityManager,
+          );
+        if (!updated) {
+          // If no rows were update it means that the data is already updated and the worker was already executed before.
+          jobLogger.log(
+            "Assessment status and workflow data were already updated. This indicates that the worker " +
+              "was invoked multiple times and it was already executed with success.",
+          );
+          return job.complete();
```

Comment:
What if line 259 was executed and then line 273 was waiting for long time and timed out.

in that case, the worker will be assigned 2nd job which will not created impacted application re-assessment and first job gets timed out as well.